src/Strategy/Cocoapods/PodfileLock.hs

```diff
@@ -23,89 +28,168 @@ import Data.HashMap.Lazy qualified as HashMap (toList)
+        traverse_ (direct . PodfilePkg . depName) lockDeps
+        -- Add transitive graph.
+        traverse_ addSpec lockPods
 
-    addSection :: Has PodfileGrapher sig m => Section -> m ()
-    addSection (DependencySection deps) = traverse_ (direct . PodfilePkg . depName) deps
-    addSection (PodSection pods) = traverse_ addSpec pods
+    toDependency :: Map Text ExternalSource -> PodfilePkg -> Set PodfileLabel -> Dependency
+    toDependency externals PodfilePkg{pkgName} = foldr applyLabel start
+      where
+        start :: Dependency
+        start =
+          Dependency
+            { dependencyType = depType
+            , dependencyName = depName
+            , dependencyVersion = Nothing
+            , dependencyLocations = []
+            , dependencyEnvironments = mempty
+            , dependencyTags = Map.empty
+            }
+
+        depType :: DepType
+        depType = case Map.lookup pkgName externals of
+          Just (ExternalGitType _) -> GitType
+          _ -> PodType
+
+        depName :: Text
+        depName = case Map.lookup pkgName externals of
+          Just (ExternalGitType ExternalGitSource{urlOf}) -> urlOf
+          _ -> pkgName
+
+        applyLabel :: PodfileLabel -> Dependency -> Dependency
+        applyLabel (PodfileVersion ver) dep =
+          dep
+            { dependencyVersion =
+                CEq <$> case Map.lookup pkgName externals of
+                  Just (ExternalGitType ExternalGitSource{tagOf, commitOf, branchOf}) -> asum [tagOf, commitOf, branchOf]
+                  _ -> Just ver
+            }
 
     addSpec :: Has PodfileGrapher sig m => Pod -> m ()
-    addSpec pod = do
-      let pkg = PodfilePkg (podName pod)
-      let pkgVersion = case Map.lookup (podName pod) (lockExternalSources lockContent) of
-            Just (ExternalGitType src) ->
-              asum
-                [ tagOf src
-                , commitOf src
-                , branchOf src
-                ]
-            _ -> Just $ podVersion pod
-
-      -- add edges between spec and specdeps
-      traverse_ (edge pkg . PodfilePkg . depName) $ podSpecs pod
-      -- add a label for version
-      label pkg $ PodfileVersion pkgVersion
+    addSpec Pod{podName, podVersion, podDeps} = do
+      -- Add edges to transitive dependencies.
+      traverse_ (edge pkg . PodfilePkg . depName) podDeps
+      -- Label pods with versions.
+      label pkg $ PodfileVersion podVersion
+      where
+        pkg = PodfilePkg podName
+
+-- We use this cache to avoid repeated invocations since gtraverse may revisit
+-- nodes. In practice, this saves quite a bit of time (e.g. on example projects,
+-- it has historically turned ~2 minute analyses into ~15 seconds).
+type PodSpecCache = Map Text PodSpecJSON
+
+buildGraph ::
+  (Has ReadFS sig m, Has Exec sig m, Has Diagnostics sig m) =>
+  Path Abs File ->
+  PodLock ->
+  m (Graphing Dependency)
+buildGraph lockFilePath lockFile@PodLock{lockExternalSources} =
```

Comment:
[Nit] Prefer if we fail early here if the user cannot execute `pod --version`, or the version is below `0.29.0`(https://github.com/CocoaPods/CocoaPods/blob/master/CHANGELOG.md#0290-2013-12-25).